Lib/glyphsLib/glyphdata.py

```diff
@@ -31,69 +31,224 @@
+    # character.
+    production_name = _lookup_production_name(glyph_name)
+
+    # Some Glyphs files use production names instead of Glyph's "nice names".
     # We catch this here, so that we can return the same properties as if
     # the Glyphs file had been following the Glyphs naming conventions.
     # https://github.com/googlei18n/glyphsLib/issues/232
-    if prodname is None:
-        rev_prodname = data.PRODUCTION_NAMES_REVERSED.get(name)
+    if production_name is None:
+        rev_prodname = data.PRODUCTION_NAMES_REVERSED.get(base_name)
         if rev_prodname is not None:
-            prodname = name
-            name = rev_prodname
-    if prodname is None:
-        prodname = name
-    unistr = data.IRREGULAR_UNICODE_STRINGS.get(name)
-    if unistr is None:
-        unistr = agl.toUnicode(prodname)
-    if unistr != "" and name not in data.MISSING_UNICODE_STRINGS:
-        unistr_result = unistr
+            production_name = base_name
+            base_name = rev_prodname
+
+    # Finally, if we couldn't find a known production name one way or another,
+    # conclude that the glyph name doesn't carry any Unicode semantics. Use the
+    # bare name in that case.
+    if production_name is None:
+        production_name = glyph_name
+
+    # Next, derive the actual character from the production name, e.g.
+    # "uni0414" -> "Д". Two caveats:
+    # 1. For some glyphs, Glyphs does not have a mapped character even when one 
+    #    could be derived.
+    # 2. For some others, Glyphs has a different idea than the agl module.
+    characters = None
+    if base_name not in data.MISSING_UNICODE_STRINGS:  # 1.
+        characters = data.IRREGULAR_UNICODE_STRINGS.get(base_name)  # 2.
+        if characters is None:
+            characters = agl.toUnicode(production_name) or None
+
+    # Lastly, generate the category in the sense of Glyph's
+    # GSGlyphInfo.category and .subCategory.
+    category, sub_category = _get_category(base_name, characters, data)
+
+    return Glyph(glyph_name, production_name, characters, category, sub_category)
+
+
+def _lookup_production_name(glyph_name, data=glyphdata_generated):
+    """Return the production name for a glyph name from the GlyphsData.xml
+    database according to the AGL specification.
+
+    Handles single glyphs (e.g. "brevecomb") and ligatures (e.g.
+    "brevecomb_acutecomb"). Returns None when a valid and semantically
+    meaningful production name can't be constructed or when the AGL
+    specification would be violated, get_glyph() will use the bare glyph
+    name then.
+
+    Note:
+    - Glyph name is the full name, e.g. "brevecomb_acutecomb.case".
+    - Base name is the base part, e.g. "brevecomb_acutecomb"
+    - Suffix is e.g. "case".
+    """
+
+    def is_unicode_u_value(name):
+        return name.startswith("u") and all(
+            part_char in "0123456789ABCDEF" for part_char in name[1:]
```

Comment:
use a set of chars instead of a string literal here maybe

Reply:
```
In [5]:  hex_characters = {
   ...:    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F",
   ...:  }

In [10]: %timeit all(part_char in hex_characters for part_char in uname[1:])
623 ns ± 3.89 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [11]: %timeit all(part_char in "0123456789ABCDEF" for part_char in uname[1:])
567 ns ± 3.34 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
Let's keep the string literal :)